src/unix/fs.c

```diff
@@ -235,6 +235,93 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
 }
 
 
+static int uv__fs_mkstemp(uv_fs_t* req) {
+  int r;
+#ifdef UV__O_CLOEXEC
+  int (*mkostemp_function)(char*, int);
+  static int no_cloexec_support;
+#endif
+  static const char pattern[] = "XXXXXX";
+  char* path;
+  size_t path_length;
+
+  path = (char*) req->path;
+  path_length = strlen(path);
+
+  /* EINVAL can be returned for 2 reasons:
+       1. The template's last 6 characters were not XXXXXX
+       2. open() didn't support O_CLOEXEC
+     We want to avoid going to the fallback path in case
+     of 1., so it's manually checked before. */
+  if (path_length < strlen(pattern) ||
+      strcmp(path + (path_length - (sizeof(pattern) - 1)), pattern)) {
+    errno = EINVAL;
+    return -1;
+  }
+
+#ifdef UV__O_CLOEXEC
+  if (no_cloexec_support == 0) {
+    *(int**)(&mkostemp_function) = dlsym(RTLD_DEFAULT, "mkostemp64");
+
+    /* We don't care about errors, but we do want to clean them up.
+       If there has been no error, then dlerror() will just return
+       NULL. */
+    dlerror();
+
+    /* If mkostemp64() doesn't exist, try mkostemp(). */
+    if (mkostemp_function == NULL) {
+      *(int**)(&mkostemp_function) = dlsym(RTLD_DEFAULT, "mkostemp");
+      dlerror();
+    }
+
+    if (mkostemp_function != NULL) {
+#ifdef O_LARGEFILE
+      r = mkostemp_function(path, UV__O_CLOEXEC | O_LARGEFILE);
+#else
+      /* FreeBSD doesn't support O_LARGEFILE. */
+      r = mkostemp_function(path, UV__O_CLOEXEC);
+#endif
+
+      if (r >= 0)
+        return r;
+
+      /* If mkostemp() returns EINVAL, it means the kernel doesn't
+         support O_CLOEXEC, so we just fallback to mkstemp() below. */
+      if (r == -1 && errno != EINVAL)
+        return r;
+
+      /* We set the static variable so that next calls don't even
+         try to use mkostemp. */
+      no_cloexec_support = 1;
+    }
+  }
+#endif
+
+  r = mkstemp(path);
+
+  if (r == -1)
+    return -1;
+
+  if (req->cb != NULL)
+    uv_rwlock_rdlock(&req->loop->cloexec_lock);
+
+  /* In case of failure `uv__cloexec` will leave error in `errno`,
+   * so it is enough to just set `r` to `-1`.
+   */
+  if (r >= 0 && uv__cloexec(r, 1) != 0) {
+    r = uv__close(r);
+    if (r != 0 && r != -EINPROGRESS)
```

Comment:
remove the EINPROGRESS check, we fixed that by adding the check inside uv__close.